nilearn/tests/test_masking.py

```diff
@@ -445,126 +544,219 @@ def test_intersect_masks():
     # +---+---+---+---+
     # |   |   | X |   |
     # +---+---+---+---+
+    mask_c = np.zeros(SHAPE_MASK, dtype=bool)
+    mask_c[:, 2] = 1
+    mask_c[0, 0] = 1
+    return mask_c
+
+
+@pytest.fixture
+def mask_c():
+    return make_mask_c()
+
+
+@pytest.fixture
+def mask_c_img():
+    return Nifti1Image(make_mask_c().astype("int32"), AFFINE_EYE)
+
+
+def make_mask_d():
+    mask_d = np.zeros((8, 8, 1), dtype=bool)
+    mask_d[2:6, 2:6] = 1
+    return mask_d
+
+
+@pytest.fixture
+def mask_d():
+    return make_mask_c()
+
+
+@pytest.fixture
+def mask_d_img():
+    return Nifti1Image(make_mask_d().astype("uint8"), AFFINE_EYE / 2.0)
+
+
+def test_intersect_masks_filename(mask_a_img, mask_b_img, expected_mask_ab):
+    with write_tmp_imgs(
+        mask_a_img, mask_b_img, create_files=True
+    ) as filenames:
+        mask_ab = intersect_masks(filenames, threshold=1.0)
+
+        assert_array_equal(get_data(mask_ab), expected_mask_ab)
+
+
+def test_intersect_masks_2_images(
+    mask_a_img,
+    mask_b_img,
+    expected_mask_ab,
+):
+    """Test the intersect_masks function"""
+    mask_ab = intersect_masks([mask_a_img, mask_b_img], threshold=1.0)
+
+    assert_array_equal(get_data(mask_ab), expected_mask_ab)
+
+
+@pytest.mark.parametrize("threshold", [0.5, 1.0])
+def test_intersect_masks_3_images(
+    threshold,
+    mask_a_img,
+    mask_b_img,
+    mask_c_img,
+    expected_mask_ab,
+):
+    """Test the intersect_masks function
+
+    default threshold is 0.5
+    """
+    mask_abc = intersect_masks(
+        [mask_a_img, mask_b_img, mask_c_img], threshold=threshold
+    )
+
+    expected_mask_abc = expected_mask_ab
+    if threshold == 0.5:
+        expected_mask_abc[1, 2] = 1
+        expected_mask_abc[3, 2] = 1
+    assert_array_equal(get_data(mask_abc), expected_mask_abc)
+
+
+@pytest.mark.parametrize("connected", [True, False])
+def test_intersect_masks_3_images_connected(
+    mask_a, mask_a_img, mask_b, mask_b_img, mask_c, mask_c_img, connected
+):
+    """Test the intersect_masks function"""
+    mask_abc = intersect_masks(
+        [mask_a_img, mask_b_img, mask_c_img],
+        threshold=0.0,
+        connected=connected,
+    )
+
+    expected_mask_abc = mask_a + mask_b + mask_c
+    if connected:
+        expected_mask_abc[0, 0] = 0
+    assert_array_equal(get_data(mask_abc), expected_mask_abc)
+
+
+def test_intersect_masks_with_f8(mask_a_img, mask_b_img, expected_mask_ab):
+    """Test intersect mask images with '>f8'.
+
+    This function uses
+    largest_connected_component to check if intersect_masks passes with
+    connected=True (which is by default)
+    """
+    mask_a_img_change_dtype = Nifti1Image(
+        get_data(mask_a_img).astype(">f8"), affine=mask_a_img.affine
+    )
+    mask_b_img_change_dtype = Nifti1Image(
+        get_data(mask_b_img).astype(">f8"), affine=mask_b_img.affine
+    )
+    mask_ab_change_type = intersect_masks(
+        [mask_a_img_change_dtype, mask_b_img_change_dtype], threshold=1.0
```

Comment:
Maybe we can remove the lines above and factorize a bit :

```suggestion
        [Nifti1Image(get_data(mask).astype(">8"), affine=mask.affine) for mask in (mask_a_img, mask_b_img)], threshold=1.0
```

Reply:
keeping the comprehension the list comprehension but will keep the intermediate variable to make it a bit easier to parse for humans